### PR TITLE
[FLINK-15191][connectors / kafka]Fix can't create table source for Kafka if watermark or computed column is defined.

### PR DESCRIPTION
**Pull request by leonardBang**
[FLINK-15191][connectors / kafka]Fix can't create table source for Kafka if watermark or computed column is defined.

## What is the purpose of the change

* This pull request Fix can't create table source for Kafka if watermark or computed column is defined. 

## Brief change log

  - *1. Add validation logic for watermark and computed column in `SchemaValidator`.*
  - *2. Add schema key support in connecter factories.*
  - *3. Build ITCase from Kafka for watermark and computed column feature.*

## Verifying this change
 - *Add IT tests in org/apache/flink/streaming/connectors/kafka/KafkaTableSourceSinkFactoryTestBase.java, it will run :
`Kafka010TableSourceSinkFactoryTest`
`Kafka011TableSourceSinkFactoryTest`
`Kafka08TableSourceSinkFactoryTest`
`Kafka09TableSourceSinkFactoryTest`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.flink.util.Preconditions;
+
+/**
+ * Utilities to filter generated column from {@link TableSchema}.
+ */
+@Internal
+public class TableSchemaUtils {
+
+	public static TableSchema filterGeneratedColumn(TableSchema tableSchema) {
+		Preconditions.checkNotNull(tableSchema);
```
Comment (danny0405) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: We may need to add some document for this method, do you think the name `filterOutGeneratedColumns` is better ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala [hunk 2]
```diff
@@ -122,7 +123,9 @@ object MemoryTableSourceSinkUtil {
     var fieldTypes: Array[TypeInformation[_]] = _
 
     override def getOutputType: TypeInformation[Row] = {
-      new RowTypeInfo(getTableSchema.getFieldTypes, getTableSchema.getFieldNames)
+      new RowTypeInfo(
+        TableSchemaUtils.filterGeneratedColumn(getTableSchema).getFieldTypes,
+          TableSchemaUtils.filterGeneratedColumn(getTableSchema).getFieldNames)
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala]: Reuse the generated schema.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/InMemoryTableFactory.scala [hunk 3]
```diff
@@ -19,19 +19,20 @@
...
 import org.apache.flink.table.descriptors.{DescriptorProperties, SchemaValidator}
+import org.apache.flink.table.descriptors.DescriptorProperties.{TABLE_SCHEMA_EXPR, WATERMARK, WATERMARK_ROWTIME, WATERMARK_STRATEGY_DATA_TYPE, WATERMARK_STRATEGY_EXPR}
 import org.apache.flink.table.factories.{StreamTableSinkFactory, StreamTableSourceFactory, TableFactory}
 import org.apache.flink.table.sinks.StreamTableSink
 import org.apache.flink.table.sources.StreamTableSource
 import org.apache.flink.table.types.logical.LogicalTypeRoot
 import org.apache.flink.table.types.utils.TypeConversions.fromDataTypeToLegacyInfo
 import org.apache.flink.types.Row
-
 import java.sql.Timestamp
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/InMemoryTableFactory.scala]: unnecessary change.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/MemoryTableSourceSinkUtil.scala [hunk 4]
```diff
@@ -103,7 +103,9 @@ object MemoryTableSourceSinkUtil {
     with AppendStreamTableSink[Row] {
 
     override def getOutputType: TypeInformation[Row] = {
-      new RowTypeInfo(getTableSchema.getFieldTypes, getTableSchema.getFieldNames)
+      new RowTypeInfo(
+        TableSchemaUtils.filterGeneratedColumn(getTableSchema).getFieldTypes,
+        TableSchemaUtils.filterGeneratedColumn(getTableSchema).getFieldNames)
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/MemoryTableSourceSinkUtil.scala]: Reuse the generated schema.

**Comment by wuchong** [comment]
And could you add some IT cases for Kafka connector? 
The DDL for kafka can contains watermark and computed column.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java (hunk 1 above)
Comment (danny0405) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: I'm actually +1 to add an interface in `TableSchema` to handle this, but because TableSchema is a public API, so i'm indecisive about that.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 5]
```diff
@@ -118,6 +131,36 @@ else if (proctimeFound) {
...
+	private void validateWatermark(DescriptorProperties properties) {
+		final String schemaWatermarkKey = SCHEMA + "." + WATERMARK;
+		int watermarkRowtimeKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_ROWTIME).size();
+		int watermarkStrategyKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_STRATEGY_EXPR).size();
+		int watermarkStrategyDataTypeKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_STRATEGY_DATA_TYPE).size();
+
+		if (watermarkRowtimeKeys > 0 || watermarkStrategyKeys > 0 || watermarkStrategyDataTypeKeys > 0) {
+			if (!isStreamEnvironment) {
+				throw new ValidationException(
+					format("Property '%s' is not allowed in a batch environment.", schemaWatermarkKey));
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: ok

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 6]
```diff
@@ -89,6 +95,13 @@ public void validate(DescriptorProperties properties) {
 			properties.validateString(SCHEMA + "." + i + "." + SCHEMA_NAME, false, 1);
 			properties.validateDataType(SCHEMA + "." + i + "." + SCHEMA_DATA_TYPE, SCHEMA + "." + i + "." + SCHEMA_TYPE, false);
 			properties.validateString(SCHEMA + "." + i + "." + SCHEMA_FROM, true, 1);
+
+			// validate generated column expression
+			final String generatedColumnExpressionKey = SCHEMA + "." + i + "." + TABLE_SCHEMA_EXPR;
+			if (properties.containsKey(generatedColumnExpressionKey)) {
+				properties.validateString(generatedColumnExpressionKey, false, 1);
+			}
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: ok

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 7]
```diff
@@ -118,6 +131,36 @@ else if (proctimeFound) {
...
+	}
+
+	/**
+	 * Validate watermarks if exists.
+	 */
+	private void validateWatermark(DescriptorProperties properties) {
+		final String schemaWatermarkKey = SCHEMA + "." + WATERMARK;
+		int watermarkRowtimeKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_ROWTIME).size();
+		int watermarkStrategyKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_STRATEGY_EXPR).size();
+		int watermarkStrategyDataTypeKeys = properties.getIndexedProperty(schemaWatermarkKey, WATERMARK_STRATEGY_DATA_TYPE).size();
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: remove them according to Jark's comment

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 8]
```diff
@@ -269,12 +320,14 @@ else if (isRowtime) {
...
-						.getOptionalBoolean(SCHEMA + "." + i + "." + SCHEMA_PROCTIME)
-						.orElse(false);
+					.getOptionalBoolean(SCHEMA + "." + i + "." + SCHEMA_PROCTIME)
+					.orElse(false);
 				boolean isRowtime = properties
-						.containsKey(SCHEMA + "." + i + "." + ROWTIME_TIMESTAMPS_TYPE);
+					.containsKey(SCHEMA + "." + i + "." + ROWTIME_TIMESTAMPS_TYPE);
+				boolean isGeneratedColumn = properties
+					.containsKey(SCHEMA + "." + i + "." + TABLE_SCHEMA_EXPR);
 				// remove proctime/rowtime from mapping
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: ok,  add it

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java (hunk 1 above)
Comment (leonardBang) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: I think `getPhysicalSchema` is a nice choice

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/InMemoryTableFactory.scala (hunk 3 above)
Comment (leonardBang) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/utils/InMemoryTableFactory.scala]: got it

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/TableSourceValidation.java [hunk 9]
```diff
@@ -59,7 +60,7 @@
 	 * @param tableSource The {@link TableSource} for which the time attributes are checked.
 	 */
 	public static void validateTableSource(TableSource<?> tableSource){
-		TableSchema schema = tableSource.getTableSchema();
+		TableSchema schema = TableSchemaUtils.filterGeneratedColumn(tableSource.getTableSchema());
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/TableSourceValidation.java]: add validation here

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.flink.table.api.TableColumn;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.table.sources.TableSource;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * Utilities to deal physical {@link TableColumn} from {@link TableSchema}.
+ */
+@Internal
```
Comment (danny0405) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: You create this class as a `TableSchema` tools right ? So change to "Utilities to {@link TableSchema}" is okey.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+public class TableSchemaUtils {
+
+	/**
+	 * Return {@link TableSchema} which consist of all physical columns.
+	 * Constructor of {@link TableSource} and {@link TableSink} should call this method.
+	 */
```
Comment (danny0405) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: Returns {@link TableSchema} which consists of all physical columns. That means, the computed columns are filtered out.

<p>Readers(or writers) such as {@link TableSource} and {@link TableSink} should always call this method.

**Comment by leonardBang** [comment]
rebase master for fixing conflict.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java [hunk 12]
```diff
@@ -118,6 +128,33 @@ else if (proctimeFound) {
...
 			}
 		}
+
+		validateWatermark(properties);
+	}
+
+	/**
+	 * Validate watermarks if exists.
+	 */
+	private void validateWatermark(DescriptorProperties properties) {
```
Comment (wuchong) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java]: I just remembered that  we don't need to validate watermark in connector side. The watermark properteis and computed column properties will be validated and parsed by planner. So we can just remove them from the validation. The only thing that connectors need to do is adding them to the supported list.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/api/TableUtils.java [hunk 13]
```diff
@@ -100,7 +101,7 @@
 	 * for external output.
 	 */
 	private static TableSchema buildNewTableSchema(Table table) {
-		TableSchema oldSchema = table.getSchema();
+		TableSchema oldSchema = TableSchemaUtils.getPhysicalTableSchema(table.getSchema());
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/api/TableUtils.java]: I think we shouldn't change this line. The schema of `Table` may contains generated columns, but the generated column should also be outputed.

**Comment by leonardBang** [comment]
@Dany @wuchong I update the PR，could you have a more look， thanks very much?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala [hunk 14]
```diff
@@ -163,4 +164,59 @@ class SchemaValidatorTest {
...
+      .field("f1", DataTypes.STRING)
+      .field("f2", DataTypes.INT)
+      .field("f3", DataTypes.TIMESTAMP(3))
+      .build()
+    val schema = SchemaValidator.deriveTableSinkSchema(properties)
+    assertEquals(expectd, schema)
+   }
+
+  @Test(expected = classOf[IllegalStateException])
+  def testSchemaWithMultiWatermark(): Unit = {
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/SchemaValidatorTest.scala]: I think we can remove this test, because multiple watermark validatation has been covered by the framework (i.e. TableSchema).

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+					builder.field(tableColumn.getName(), tableColumn.getType());
+				}
+			});
+		return builder.build();
+	}
+
+	/**
+	 * Returns whether there contains the generated {@link TableColumn} such as computed column and watermark.
+	 */
+	public static boolean containsGeneratedColumn(TableSchema tableSchema) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java]: Remove unused method.
